File: src/java/main/src/main/java/com/tencent/wcdb/winq/MultiTypeArray.java
// Created by qiuwenchen on 2023/4/14.
//

/*
 * Tencent is pleased to support the open source community by making
 * WCDB available.
 *
 * Copyright (C) 2017 THL A29 Limited, a Tencent company.
 * All rights reserved.
 *
 * Licensed under the BSD 3-Clause License (the "License"); you may not use
 * this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *       https://opensource.org/licenses/BSD-3-Clause
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.tencent.wcdb.winq;

import com.tencent.wcdb.base.CppObject;
import com.tencent.wcdb.base.Value;

import java.util.Arrays;

class ObjectType {
    final static int Null = 0;
    final static int Bool = 1;
    final static int Char = 2;
    final static int Byte = 3;
    final static int Short = 4;
    final static int Int = 5;
    final static int Long = 6;
    final static int Float = 7;
    final static int Double = 8;
    final static int String = 9;
    final static int Identifier = 10;

    final static int Value = 11;
    final static int Unknown = 12;
}

class MultiTypeArray {

    int[] types;
    long[] longValues;
    double[] doubleValues;
    String[] stringValues;

    MultiTypeArray(Object[] values) {
        int valueCount = values.length;
        types = new int[valueCount];
        longValues = new long[valueCount];
        doubleValues = new double[valueCount];
        stringValues = new String[valueCount];
        int longIndex = 0;
        int doubleIndex = 0;
        int stringIndex = 0;

        for (int i = 0; i < valueCount; i++) {
            Object obj = values[i];
            int valueType = getObjectType(obj);
            switch (valueType) {
                case ObjectType.Null:
                    types[i] = CPPType.Null;
                    longIndex++;
                    break;
                case ObjectType.Bool:
                    types[i] = CPPType.Bool;
                    longValues[longIndex] = (Boolean) obj ? 1 : 0;
                    longIndex++;
                    break;
                case ObjectType.Byte:
                    types[i] = CPPType.Int;
                    longValues[longIndex] = (Byte) obj;
                    longIndex++;
                    break;
                case ObjectType.Char:
                    types[i] = CPPType.Int;
                    longValues[longIndex] = (Character) obj;
                    longIndex++;
                    break;
                case ObjectType.Short:
                    types[i] = CPPType.Int;
                    longValues[longIndex] = (Short) obj;
                    longIndex++;
                    break;
                case ObjectType.Int:
                    types[i] = CPPType.Int;
                    longValues[longIndex] = (Integer) obj;
                    longIndex++;
                    break;
                case ObjectType.Long:
                    types[i] = CPPType.Int;
                    longValues[longIndex] = (Long) obj;
                    longIndex++;
                    break;
                case ObjectType.Float:
                    types[i] = CPPType.Double;
                    doubleValues[doubleIndex] = (Float) obj;
                    doubleIndex++;
                    break;
                case ObjectType.Double:
                    types[i] = CPPType.Double;
                    doubleValues[doubleIndex] = (Double) obj;
                    doubleIndex++;
                    break;
                case ObjectType.String:
                    types[i] = CPPType.String;
                    stringValues[stringIndex] = (String) obj;
                    stringIndex++;
                    break;
                case ObjectType.Identifier:
                    Identifier identifier = (Identifier) obj;
                    types[i] = Identifier.getCppType(identifier);
                    longValues[longIndex] = CppObject.get(identifier);
                    longIndex++;
                    break;
                case ObjectType.Value:
                    Value valueObj = (Value) obj;
                    switch (valueObj.getType()) {
                        case Null:
                            types[i] = CPPType.Null;
                            longIndex++;
                            break;
                        case Integer:
                            types[i] = CPPType.Int;
                            longValues[longIndex] = valueObj.getLong();
                            longIndex++;
                            break;
                        case Float:
                            types[i] = CPPType.Double;
                            doubleValues[doubleIndex] = valueObj.getDouble();
                            doubleIndex++;
                            break;
                        case Text:
                            types[i] = CPPType.String;
                            stringValues[stringIndex] = valueObj.getText();
                            stringIndex++;
                            break;
                    }
                    break;
            }
        }
        if (stringValues.length * 0.75 > stringIndex) {
            if (stringIndex == 0) {
                stringValues = null;
            } else {
                stringValues = Arrays.copyOf(stringValues, stringIndex);
            }
        }
    }

    protected static int getObjectType(Object obj) {
        if (obj == null) {
            return ObjectType.Null;
        } else if (obj instanceof Identifier) {
            return ObjectType.Identifier;
        }
        Class cls = obj.getClass();
        if (cls == String.class) {
            return ObjectType.String;
        } else if (cls == Integer.class) {
            return ObjectType.Int;
        } else if (cls == Float.class) {
            return ObjectType.Float;
        } else if (cls == Double.class) {
            return ObjectType.Double;
        } else if (cls == Boolean.class) {
            return ObjectType.Bool;
        } else if (cls == Short.class) {
            return ObjectType.Short;
        } else if (cls == Long.class) {
            return ObjectType.Long;
        } else if (cls == Character.class) {
            return ObjectType.Char;
        } else if (cls == Byte.class) {
            return ObjectType.Byte;
        } else if (cls == Value.class) {
            return ObjectType.Value;
        }
        return ObjectType.Unknown;
    }
}

File: src/rust/wcdb/src/winq/multi_type_array.rs
use crate::base::cpp_object::CppObject;
use crate::base::value::Value;
use crate::winq::column_type::ColumnType;
use crate::winq::identifier::{CPPType, Identifier};
use crate::winq::object::Object;
use std::any::Any;
use std::ffi::c_double;

#[repr(i32)]
pub enum ObjectType {
    Null,
    Bool,
    Char,
    Byte,
    Short,
    Int,
    Long = 6,
    Float,
    Double,
    String,
    Identifier,
    Value,
    Unknown,
}

pub struct MultiTypeArray {
    pub(crate) types: Vec<i32>,
    pub(crate) long_values: Vec<i64>,
    pub(crate) double_values: Vec<c_double>,
    pub(crate) string_values: Option<Vec<String>>,
}

impl MultiTypeArray {
    pub fn new_with_objects(values: &Vec<Object>) -> Self {
        let value_count = values.len();

        let mut types: Vec<i32> = vec![0; value_count];
        let mut long_values: Vec<i64> = vec![0; value_count];
        let mut double_values = vec![0.0; value_count];
        let mut string_values = vec![String::new(); value_count];

        let mut long_index = 0;
        let mut double_index = 0;
        let mut string_index = 0;

        for (i, obj) in values.iter().enumerate() {
            match obj {
                Object::Null => {
                    types[i] = CPPType::Null as i32;
                    long_index += 1;
                }
                Object::Bool(b) => {
                    types[i] = CPPType::Bool as i32;
                    long_values[long_index] = if *b { 1 } else { 0 } as i64;
                    long_index += 1;
                }
                Object::Byte(b) => {
                    types[i] = CPPType::Int as i32;
                    long_values[long_index] = *b as i64;
                    long_index += 1;
                }
                Object::Char(c) => {
                    types[i] = CPPType::Int as i32;
                    long_values[long_index] = *c as i64;
                    long_index += 1;
                }
                Object::Short(s) => {
                    types[i] = CPPType::Int as i32;
                    long_values[long_index] = *s as i64;
                    long_index += 1;
                }
                Object::Int(int) => {
                    types[i] = CPPType::Int as i32;
                    long_values[long_index] = *int as i64;
                    long_index += 1;
                }
                Object::Long(l) => {
                    types[i] = CPPType::Int as i32;
                    long_values[long_index] = *l as i64;
                    long_index += 1;
                }
                Object::Float(f) => {
                    types[i] = CPPType::Double as i32;
                    double_values[double_index] = *f as c_double;
                    double_index += 1;
                }
                Object::Double(d) => {
                    types[i] = CPPType::Double as i32;
                    double_values[double_index] = *d as c_double;
                    double_index += 1;
                }
                Object::String(s) => {
                    types[i] = CPPType::String as i32;
                    string_values[string_index] = s.clone();
                    string_index += 1;
                }
                Object::Identifier(identifier) => {
                    types[i] = Identifier::get_cpp_type(identifier) as i32;
                    long_values[long_index] = CppObject::get(identifier) as i64;
                    long_index += 1;
                }
                Object::Value(value_obj) => match value_obj.get_type() {
                    ColumnType::Null => {
                        types[i] = CPPType::Null as i32;
                        long_index += 1;
                    }
                    ColumnType::Integer => {
                        types[i] = CPPType::Int as i32;
                        long_values[long_index] = value_obj.get_i64() as i64;
                        long_index += 1;
                    }
                    ColumnType::Float => {
                        types[i] = CPPType::Double as i32;
                        double_values[double_index] = value_obj.get_f64() as c_double;
                        double_index += 1;
                    }
                    ColumnType::Text => {
                        types[i] = CPPType::String as i32;
                        string_values[string_index] = value_obj.get_text();
                        string_index += 1;
                    }
                    _ => {}
                },
            }
        }

        let string_values = if string_values.len() as f64 * 0.75 > string_index as f64 {
            if string_index == 0 {
                None
            } else {
                Some(string_values[0..string_index].to_vec())
            }
        } else {
            Some(string_values)
        };

        MultiTypeArray {
            types,
            long_values,
            double_values,
            string_values,
        }
    }

    pub fn get_object_type(val: Box<dyn Any>) -> ObjectType {
        if val.is::<Identifier>() {
            return ObjectType::Identifier;
        } else if val.is::<&str>() {
            return ObjectType::String;
        } else if val.is::<i32>() {
            return ObjectType::Int;
        } else if val.is::<f32>() {
            return ObjectType::Float;
        } else if val.is::<f64>() {
            return ObjectType::Double;
        } else if val.is::<bool>() {
            return ObjectType::Bool;
        } else if val.is::<i16>() {
            return ObjectType::Short;
        } else if val.is::<i64>() {
            return ObjectType::Long;
        } else if val.is::<char>() {
            return ObjectType::Char;
        } else if val.is::<i8>() {
            return ObjectType::Byte;
        } else if val.is::<Value>() {
            return ObjectType::Value;
        }
        ObjectType::Unknown
    }

    pub fn types(&self) -> &Vec<i32> {
        &self.types
    }

    pub fn long_values(&self) -> &Vec<i64> {
        &self.long_values
    }

    pub fn double_values(&self) -> &Vec<c_double> {
        &self.double_values
    }

    pub fn string_values(&self) -> &Option<Vec<String>> {
        &self.string_values
    }
}
